src/core/term_dag.rs

```diff
@@ -116,6 +115,111 @@ impl<Witness, App: Application> TermDag<Witness, App> {
     pub fn jet(jet: &'static JetNode<App>) -> Rc<Self> {
         Rc::new(TermDag::Jet(jet))
     }
+
+    /// Return the left child of the given DAG root, if there is such a child.
+    pub fn get_left(&self) -> Option<&Self> {
+        match self {
+            TermDag::InjL(l) | TermDag::InjR(l) | TermDag::Take(l) | TermDag::Drop(l) => Some(l),
+            TermDag::Comp(l, _)
+            | TermDag::Case(l, _)
+            | TermDag::Pair(l, _)
+            | TermDag::AssertL(l, _)
+            | TermDag::AssertR(l, _)
+            | TermDag::Disconnect(l, _) => Some(l),
+            _ => None,
+        }
+    }
+
+    /// Return the right child of the given DAG root, if there is such a child.
+    pub fn get_right(&self) -> Option<&Self> {
+        match self {
+            TermDag::Comp(_, r)
+            | TermDag::Case(_, r)
+            | TermDag::Pair(_, r)
+            | TermDag::AssertL(_, r)
+            | TermDag::AssertR(_, r)
+            | TermDag::Disconnect(_, r) => Some(r),
+            _ => None,
+        }
+    }
+
+    /// Visit the DAG and return a stack that can be popped such that the nodes are in post order,
+    /// i.e., children appear before their parent.
+    fn get_post_order_stack(&self) -> Vec<&TermDag<Witness, App>> {
+        let mut visited = Vec::new();
+        let mut to_visit = vec![self];
+
+        while let Some(dag) = to_visit.pop() {
+            visited.push(dag);
+
+            if let Some(l) = dag.get_left() {
+                if !l.is_cross_reference(&to_visit) {
+                    to_visit.push(l);
+                }
+            }
+            if let Some(r) = dag.get_right() {
+                if !r.is_cross_reference(&to_visit) {
+                    to_visit.push(r);
+                }
+            }
+        }
+
+        visited
+    }
+
+    /// Return whether `self` is a cross-reference to a node that appears
+    /// earlier in the post-order of the DAG.
+    ///
+    /// Using the two-stack approach of [`Self::get_post_order_stack`],
+    /// nodes that appear earlier are reachable from the nodes in `to_visit`.
+    fn is_cross_reference(&self, to_visit: &[&TermDag<Witness, App>]) -> bool {
+        for dag in to_visit {
+            if self.is_reachable_from(dag) {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    // TODO: Slow for large programs, like HASHBLOCK or SCHNORR
+    // Memoize reachable nodes? Save unique pairs of parents to leftmost child (requires additional iteration).
+    /// Return whether the DAG root `self` is reachable from the DAG root `other`.
+    ///
+    /// Performs DFS internally.
+    fn is_reachable_from(&self, other: &Self) -> bool {
+        let mut to_visit = vec![other];
+
+        while let Some(node) = to_visit.pop() {
+            if std::ptr::eq(node, self) {
```

Comment:
In 41ae92f368f222738c1b99058f11f723b86e1155:

This is a neat trick, but I worry about using pointer identity in Rust. I **believe** this code is correct because nothing will move once it's in an `Rc` [but strictly speaking this is not guaranteed and we'd need to think about pinning rules etc to be sure](https://stackoverflow.com/questions/72148631/how-can-i-hash-by-a-raw-pointer). Are we able to use CMR here instead? I guess not, we'd need to use the IMR which we don't yet have.

Reply:
I think, there is no way around pointer equality. `TermDag`s have no other kind of identity _(as of now, and we probably don't want to add CMRs to the enum)_.  I tried to define a notion of paths from root, but this approach suffers from the fact that the same shared node can simply have more than one path, depending on where you come from. @sanket1729 's [code also uses pointer equality to find shared DAGs.](https://github.com/ElementsProject/rust-simplicity/blob/master/src/core/term_dag.rs#L169)

Reply:
Yes, on further reflection I think we should stick with pointer equality here.

And even if we did want to add a CMR this is actually not a sufficient notion of "identity" for this purpose..